src/mem_writer.c

```diff
@@ -139,8 +141,106 @@ void mem_writer_i64(mem_writer* writer, int64_t num)
 
 void mem_writer_f64(mem_writer* writer, double num)
 {
-    // TODO: This is a candidate for optimisation by avoiding printf.
-    // The Ryu project sees appropriate here: https://github.com/ulfjack/ryu
-    // Specifically the `int d2s_buffered_n(double f, char* result);` function.
-    mem_writer_printf(writer, "%g", num);
+    // Note, this code will never emit exponent notation.
+    // TODO: Multi-threading locks.
+
+    // Shortest string that yields `d` when read in and rounded to nearest.
+    const int mode = 0;
+
+    // Do not limit the number of significant digits.
+    const int ndigits = 0;
+
+    // position of the decimal point in the returned result.
+    int decpt = 0;
+
+    // Set to non-zero if negative.
+    int sign = 0;
+
+    // Set to the end of the buffer.
+    char* buf_end = NULL;
+
+    // Returns buffer of integer digits or special value, no '.' chars.
+    char* buf = dtoa(num, mode, ndigits, &decpt, &sign, &buf_end);
+    assert((buf_end != NULL) && (buf_end >= buf));
+
+    const size_t len = buf_end - buf;
+    assert(len != 0);
+
+    if (buf[0] == 'N')
+    {
+        mem_writer_str(writer, 3, "NaN");
+    }
+    else if (buf[0] == 'I')
+    {
+        if (sign)
+            mem_writer_str(writer, 9, "-Infinity");
+        else
+            mem_writer_str(writer, 8, "Infinity");
+    }
+    else if (decpt <= 0)
+    {
+        char* dest = NULL;
+        const size_t leading_zeros = (size_t)(-decpt);
+        size_t book_len = 2 + leading_zeros + len;
+        if (sign)
```

Comment:
Can be converted to
```
        size_t book_len = sign + 2 + leading_zeros + len;
        dest = mem_writer_book(writer, book_len);
        if (sign) *dest++ = '-';
```
and save a few lines of code.

Reply:
Good point, almost. Really should be `size_t book_len = !!sign + 2 + leading_zeros + len;` (as sign may not be exatcly 1).

Reply:
You are right.
Anyway dtoa_r set it to 1 as was expected.